### PR TITLE
Added a SoftLayer Provider

### PR DESCRIPTION
**Pull request by jbd214**
This adds support for SoftLayer Linux benchmarks.

### CODE REVIEW

**Comment by jbd214** [comment]
A manager has signed the CLA. I am listed in the Google Group IBM-PerfKitBenchmarker-Contributors.

**Comment by noahl** [comment]
@jbd214 thanks a lot for contributing this! The CLAbot is automated - please reply "I signed it!" (even though a manager actually signed it) to trigger it to check. If there are still problems, we'll start debugging.

**Comment by jbd214** [comment]
I signed it!

**Comment by hildrum** [comment]
For Corporate CLAs, when signed, there's a google group associated with them and that group is the list of email addresses authorized to contribute on behalf of that corporation.  The commits have to be authored with an email address in the google group associated with the Corporate CLA.

It can take a few days for a Corporate CLA to be recognized; I do see a Corporate CLA from IBM has been accepted, but I don't know if yours is one of those, or a new one.

Assuming the CLA is accepted and that's all taken care of, for the googlebot to see that the CLA is signed, you need to amend the commits so that they use an email address from the google group.  Once you've done that, you can push (probably with -f) the amended commits.  Let me know if you need more specific directions.

**Comment by MHBauer** [comment]
Will probably need to have `mwalli` in the CLA as well. `tedhtchang` should be added to/covered by the same CLA if possible.

**Comment by duglin** [comment]
@jbd214 can you please ping me dug at us.ibm.com

File: README.md [hunk 1]
```diff
@@ -535,6 +537,19 @@ You can find more details here: https://developer.rackspace.com/docs/rack-cli/co
 **Note:** Not all flavors are supported on every region. Always check first
 if the flavor is supported in the region.
 
+### Installing CLIs and credentials for SoftLayer
+
+Make sure you have installed pip (see the section above).
+
+run the following command (omit the 'sudo' on Windows)
```
Comment (hildrum) [inline README.md]: Capitalize "Run"?

File: perfkitbenchmarker/configs/default_config_constants.yaml [hunk 2]
```diff
@@ -14,6 +14,10 @@ default_single_core: &default_single_core
     machine_type: m3.medium
     zone: us-east-1
     image: null
+  SoftLayer:
+    machine_type: '{"cpus": 4, "memory": 4096, "os": "UBUNTU_LATEST_64", "nic": 1000}'
```
Comment (hildrum) [inline perfkitbenchmarker/configs/default_config_constants.yaml]: It looks like you're using a four core machine as the `default_single_core` for Softlayer.   Is this intentional?

File: perfkitbenchmarker/providers/__init__.py [hunk 3]
```diff
@@ -31,69 +31,67 @@
...
-
-  Args:
-    providers: series of strings. Each element is a value from VALID_CLOUDS
-        indicating a cloud provider for which to import the flags module.
-  """
-  for provider_name in providers:
-    normalized_name = provider_name.lower()
-    flags_module_name = '.'.join((__name__, normalized_name, 'flags'))
-    importlib.import_module(flags_module_name)
+    """Imports just the flags module for each provider.
```
Comment (hildrum) [inline perfkitbenchmarker/providers/__init__.py]: Sorry to be nit-picky, but we usually have a blank line between the one-line summary and the longer description, and between the longer description and the args.  

Also, we typically use two spaces for the indent. 

We generally follow the [google python style guide](https://google.github.io/styleguide/pyguide.html)

File: perfkitbenchmarker/providers/softlayer/__init__.py [hunk 4]
```diff
@@ -0,0 +1,13 @@
+# Copyright 2014 PerfKitBenchmarker Authors. All rights reserved.
```
Comment (hildrum) [inline perfkitbenchmarker/providers/softlayer/__init__.py]: 2016?

File: perfkitbenchmarker/providers/softlayer/flags.py [hunk 5]
```diff
@@ -0,0 +1,19 @@
+# Copyright 2015 PerfKitBenchmarker Authors. All rights reserved.
```
Comment (hildrum) [inline perfkitbenchmarker/providers/softlayer/flags.py]: 2016?

File: perfkitbenchmarker/providers/softlayer/provider_info.py [hunk 6]
```diff
@@ -0,0 +1,26 @@
+# Copyright 2015 PerfKitBenchmarker Authors. All rights reserved.
```
Comment (hildrum) [inline perfkitbenchmarker/providers/softlayer/provider_info.py]: 2016?

File: perfkitbenchmarker/providers/softlayer/softlayer_disk.py [hunk 7]
```diff
@@ -0,0 +1,168 @@
+# Copyright 2014 PerfKitBenchmarker Authors. All rights reserved.
```
Comment (hildrum) [inline perfkitbenchmarker/providers/softlayer/softlayer_disk.py]: 2016?

File: perfkitbenchmarker/providers/softlayer/softlayer_virtual_machine.py [hunk 8]
```diff
@@ -0,0 +1,472 @@
+#
```
Comment (hildrum) [inline perfkitbenchmarker/providers/softlayer/softlayer_virtual_machine.py]: Copyright line?

File: perfkitbenchmarker/providers/softlayer/softlayer_virtual_machine.py [hunk 9]
```diff
@@ -0,0 +1,472 @@
...
+    return json.dumps(mappings)
+
+
+class SoftLayerVirtualMachine(virtual_machine.BaseVirtualMachine):
+    """Object representing an SoftLayer Virtual Machine."""
+
+    CLOUD = providers.SOFTLAYER
+    DEFAULT_ROOT_DISK_TYPE = 'standard'
+
+    _lock = threading.Lock()
```
Comment (hildrum) [inline perfkitbenchmarker/providers/softlayer/softlayer_virtual_machine.py]: Why do you need a lock?  Can you add a comment explaining?

File: perfkitbenchmarker/providers/softlayer/softlayer_virtual_machine.py [hunk 10]
```diff
@@ -0,0 +1,472 @@
...
+
+            self.deleted_keyfile_set.add(self.zone)
+            if self.zone in self.imported_keyfile_set:
+                self.imported_keyfile_set.remove(self.zone)
+
+    @vm_util.Retry(log_errors=False)
+    def _PostCreate(self):
+        """Check the status of the system to ensure that it is active"""
+
+        """Get the instance's data and tag it."""
```
Comment (hildrum) [inline perfkitbenchmarker/providers/softlayer/softlayer_virtual_machine.py]: We usually use `# .....` for these type of comments.

File: perfkitbenchmarker/providers/softlayer/util.py [hunk 11]
```diff
@@ -0,0 +1,80 @@
+# Copyright 2014 PerfKitBenchmarker Authors. All rights reserved.
```
Comment (hildrum) [inline perfkitbenchmarker/providers/softlayer/util.py]: 2016?

File: requirements-softlayer.txt [hunk 12]
```diff
@@ -0,0 +1,19 @@
+
+# Copyright 2015 PerfKitBenchmarker Authors. All rights reserved.
```
Comment (hildrum) [inline requirements-softlayer.txt]: 2016?

**Comment by hildrum** [comment]
I took a first pass.  

I noticed you have a useful  `ReadMeSoftLayer.txt`; however, we are trying to keep providers-specific stuff to the provider directory, so I think it should be in `perfkitbenchmarker/providers/softlayer`, perhaps named just `README.txt` or `README.md` if you want to use markdown.

I notice you have a SoftLayerDisk with `_Create`, `_Delete`, and `_Exists` methods, which are all basically empty.  Perhaps you could add a comment explaining why that's the case?  Also, `_Exists` is optional, you don't have to implement it at all.  If you do implement it, always returning true is a problem, because when pkb tries to delete a disk, it first calls delete, then calls exists, and then calls delete again if Exists returns true.  

Also, stylistically, we use two spaces to indent, and I think you're using four.

**Comment by noahl** [comment]
One quick suggestion following up on what Kris said about disks. If SoftLayer machines can't add and remove disks (or if you just don't want to implement that in v1), then maybe look at mimicing GCP's "local disk" support. I believe those are statically attached to their VMs, so they should be handled very similarly.

**Comment by jbd214** [comment]
noahl, 
Thanks for the suggestion. I am refactoring the disk support now. For v1, I am targeting the simplest storage types, local & SAN disks which are part of the VS create.

**Comment by jbd214** [comment]
Kris Hildrum, 
FYI, I think I addressed your initial issues in the last commit a few days ago.

**Comment by jbd214** [comment]
Thanks for the update. I will make sure our GitHub primary emails are
listed the Google group that is associated with the CLA.

**Comment by jbd214** [comment]
The CLA that we were working on was rejected because we didn't have an attorney or executive sign it. I am closing this pull request until we find someone to sign a new CLA.

**Pull request closed without merging**